### DISCUSSION TITLE
useForm and derivatives need a clear distinction between props to be send to the server and form props

### DISCUSSION THREAD

**Discussion by vinclou** [refinedev/refine#2719] [Ideas]
## `The Problem`:

1. Workarounds around this limitation blur one's programmer intention and make code less clear in terms of type safety and just adding code bloat.
2. Question the use of utility hooks offered by Refine since u end up using 10% - 30% of the intended func.
3. Makes the utility library less flexible

## `Proposal: ` 

Because FormProps is the same as the useMutation it causes lots of hassle, very often I face a situation where I can not precisely map form values to what gets sent to the server, or for some reason don't want to or need to process updates in a particular way, before the submission.

### If as a user I can do something like this: 

And if I don't want to the default is mapped to subscription updates(it might not work out with TS)

const exp = useForm<Subscription, HttpError, SubscriptionUpdates, IFormSubscriptionUpdates>({...Props})

Then essentially I wouldn't need to use extra hooks, overrides, or other workarounds you could possible come up with to 
deal with form values being different from server-sent props.

## Let's start with an example.

```tsx
/**
* @description One of the possible workarounds to deal with type differences
*  In the example below date needs to be converted to API format.
* The ideal solution would be just to convert it with onFinish or onChange events and use the mutation returned by
* useForm hook, but in this situation, TS won't let you do it and complain.
*
* There are plenty of workarounds, so don't focus on bad code =) 
*/
function EditSubscriptionForm({ subscriptionId, termEnd }: { subscriptionId: string; termEnd: TPossibleVal<number> }) {
	const { formProps: editFormProps } = useForm<Subscription, HttpError, IFormSubscriptionUpdates>({
		action: 'edit',
		resource: `accounts/Internal/v1/Subscriptions/${subscriptionId}`,
		redirect: false,
		invalidates: ['all'],
		queryOptions: {
			enabled: false
		}
	})       
	const { mutate: editSubscription } = useUpdate<Subscription, HttpError, SubscriptionUpdates>()
	const termEndDateObj = termEnd ? CalendarDate.fromApiFormat(termEnd).toDate() : new Date()

	const handleFormUpdates = (values: IFormSubscriptionUpdates) => {
		const { currentTermEnd, quantity } = values

		if (!currentTermEnd && !quantity) {
			message.error('Please enter a new term end date or quantity, to submit a form')
			return
		}

		const subUpdates: SubscriptionUpdates = {
			...(quantity && quantity !== 0 && { quantity: quantity })
		}

		if (currentTermEnd) {
			const newTermEnd = convertToApiDate(currentTermEnd)
			subUpdates.currentTermEnd = newTermEnd
		}

		editSubscription({
			resource: `accounts/Internal/v1/Subscriptions`,
			id: subscriptionId,
			values: subUpdates,
			invalidates: ['all']
		})
	}
	return (
		<div className=''>
			<Form<IFormSubscriptionUpdates> {...editFormProps} onFinish={handleFormUpdates}>
				<Form.Item className='w-full' name='quantity' label='Quantity' initialValue={1}>
					<InputNumber min={0} max={999} />
				</Form.Item>
				<Form.Item className='w-full' name='currentTermEnd' label='Term End' initialValue={dayjs(termEndDateObj)}>
					<DatePicker
						allowClear
						id='end date picker'
						disabledDate={current => {
							const start = dayjs()
							return current < dayjs(start) || current > dayjs(termEndDateObj)
						}}
					/>
				</Form.Item>
				<Button type='primary' htmlType='submit'>
					Submit
				</Button>
			</Form>
		</div>
	)
}
```

**Comment by omeraplak** [comment]
Hey @vinclou ,
Good point, this was a problem I faced from time to time. Do you want to create a PR?

**Reply by vinclou** [reply]
Will take a look at it

**Reply by omeraplak** [reply]
Thanks! Can you create an issue if you give up? I'm sure it will be useful for hacktoberfest event 🎃

**Reply by vinclou** [reply]
Sure @omeraplak , I'll peek at it on Sunday, potentially might not even be a big change to make it possible; But I haven't looked at the code yet, too early to call it out

**Reply by MetaMmodern** [reply]
@vinclou hello, did you have any chance to check or add a fix?

**Reply by vinclou** [reply]
@MetaMmodern 
Short answer: I'm pretty loaded with work, so I have not started the implementation. 

Long: Yet, I've ended up so often overriding the onFinish function, or custom submit button behavior, that his issue slowly started fading away. Even the cost of going back and refactoring to `the desired way` would quickly add up after building a whole app with Refine.

If I ever implement this or would implement this.  I would totally do it in conjunction to adding a different route for mutations:
`Refine goes with a premise that mutation happens at the same route as fetch request` which was and is - the biggest disappointment for me in the patterns Refine supplies. As if the gated application that sees domain-restricted routes needs to modify data it always has to create a mutation hook for that. 

I've done some experiments with the custom data provider, but nothing has really solved this problem. And mutation route is pretty much the same except internal prefix, those are just inaccessible for public API, or could be used outside correct application `claim_id` 

So if you decide to work on it, I'd totally be interested to discuss the typescript part of it of sending a get_res type and mutation type.

**Reply by MetaMmodern** [reply]
Hi @vinclou . For the problem that you described with custom route for mutation: there are actually two options, maybe you missed them. One is to use `useCustomMutation` and call mutate for it. This us if you have a completely different request to make.
Another one is to modify an existing dataProvider. I've done that and refine docs have an article on how to do it. If your prefix is applied on all resources you can modify the dataprovider and inject necessary logic for routing in it. For instance, i had a problem that my API had editing resource endpoint on a PUT request, while refine was making POST request by default. So I modified data provider for all edit requests to be PUT and not POST.

However, I'm interested in solution of this discussion not because I have a different route, but because on the same resource I have different schemas for read and for update. For instance, I have an ordered list of resources, which have an "index" field each. When I create a new resource I don't want user to input the index, my backend handles this part. So in request I have to submit a different object, the one without an "index" field. That's where typescript fails. 
I've created a post on discord about this problem and Ömer told that the fix should be there in about a week. But again, this is a fix for types, not for the situation you described.
I'll give you a link once I get to PC, it has a repro stackblitz as well.

**Reply by vinclou** [reply]
@MetaMmodern Yeah, 
I have ended up doing the custom data provider for a couple of similar reasons, and never had to use useCustomMutation.

I've used bad terminology by custom mutations I meant something like this:
##### Here's an example of an ideal use case for this feature story
```tsx
        // Say we have a useForm hook, it does two things:
        // 1. Fetch the data
        // 2. Mutate the data
        // But both the mutation and fetch share the same resource prop
	const { formProps: editFormProps } = useForm<
              Subscription, 
              HttpError,
              IFormSubscriptionFormVals,
              IFormSubscriptionUpdates
          >({
		action: 'edit',
                 // Here's the GET request route
		resource: `accounts/v1/Subscriptions/${subscriptionId}`,
		// Doesn't exist in Refine, but what I propose
		mutationRes: `accounts/v1/Internal/Subscriptions/${subscriptionId}`
		redirect: false,
		invalidates: ['all'],
		queryOptions: {
			enabled: Boolean(subscriptionId)
		}
	})      
```

So in example above we pass both to props `resource` and `mutationRes` as well as what's being fetched `Subscription` and 
`IFormSubscriptionFormVals` are form vals, and `IFormSubscriptionUpdates` would be my updates.

`Yet, as mentioned the type issue went away for the most pages, as I override the onFinish for custom processing, in almost every case, I think I have like one or two pages, that are lucky enough to just use the public API and not need any overrides`

```tsx
/* 
This what the type signature might look like, but probably the core form hook needs to modified first.
As well as the core hook props to accept a `mutation resource route`
*/
export type UseFormProps<
    TData extends BaseRecord = BaseRecord,
    TError extends HttpError = HttpError,
    TFormVariables = {},
    TSubmissionVariables = {} // defaults to TFormVariables, if not provided
> = UseFormPropsCore<TData, TError, TVariables> & {
    submitOnEnter?: boolean;
    /**
     * Shows notification when unsaved changes exist
     */
    warnWhenUnsavedChanges?: boolean;
};
```

**Reply by MetaMmodern** [reply]
Okay, I see. Then we're dealing with different types of issues here. You have different endpoints for the same resource. I have different scemas/interfaces for the same resource when getting and editing the resource. Althought your issue kinda includes my issue.
We'll see what will refine core team come up with :)

**Reply by vinclou** [reply]
@MetaMmodern Well it's not an easy feature in terms of all different parts of refine pkgs to update, starting from the core lib, updating type signatures is relatively straightforward(your use case), my proposed feature requires whole lotta testing + making sure it's non-breaking, t.e if mutationRoute is not present the things get defaulted to the resource route instead.

It is a medium-risk feature, but probably the core Refine team should investigate how popular is this use case; I would argue
it adds lotta flexibility anyway and removes the boiler plate too. Since when you declare your won mutation it is essentially code
duplication, since there's a mutation declared under the hood, you are just forced to decline to use it.

**Reply by vinclou** [reply]
@omeraplak But I do think this will make ur framework more flexible and users would write less boilerplate code without the custom mutation hooks that currently need to be declared. 

`I'll probably write up a discussion with my Feedback on Refine and address the good and bad things` it might help you plan for this year's roadmap.

**Reply by aliemir** [reply]
Hey @MetaMmodern and @vinclou, we've released our roadmap for refine v4 and posted a RFC here https://github.com/refinedev/refine/discussions/3527 We'd love to hear your comments and feedbacks about it. (Not mentioned in the roadmap but we will work on this discussion too)

**Reply by vinclou** [reply]
@aliemir Thanks, will take a look over it!
